### PR TITLE
Adds confetti velocity information in draw methods

### PR DESCRIPTION
**Pull request by magneticflux-**
A redone and rebased version of #25

### CODE REVIEW

**Comment by magneticflux-** [comment]
@jinatonic I've made the changes from https://github.com/jinatonic/confetti/issues/24#issuecomment-405681688

File: confetti/src/main/java/com/github/jinatonic/confetti/confetto/Confetto.java [hunk 1]
```diff
@@ -27,39 +27,34 @@
...
+ * All of the configured states are in milliseconds, e.g. pixels per millisecond for velocity.
  */
 public abstract class Confetto {
     private static final int MAX_ALPHA = 255;
     private static final long RESET_ANIMATION_INITIAL_DELAY = -1;
 
     private final Matrix matrix = new Matrix();
     private final Paint workPaint = new Paint(Paint.ANTI_ALIAS_FLAG);
-
+    protected float currentVelocityX, currentVelocityY, currentRotationalVelocity;
```
Comment (jinatonic) [inline confetti/src/main/java/com/github/jinatonic/confetti/confetto/Confetto.java]: i would move this below definition of `currentX`

File: confetti/src/main/java/com/github/jinatonic/confetti/confetto/Confetto.java [hunk 2]
```diff
@@ -348,18 +363,15 @@ private void draw(Canvas canvas, float x, float y, float rotation, float percent
...
      *
-     * @param canvas the canvas to draw on.
-     * @param matrix an identity matrix to use for draw manipulations.
-     * @param paint the paint to perform canvas draw operations on. This paint has already been
-     *   configured via {@link #configurePaint(Paint)}.
-     * @param x the x position of the confetto relative to the canvas.
-     * @param y the y position of the confetto relative to the canvas.
-     * @param rotation the rotation (in degrees) to draw the confetto.
-     * @param percentAnimated the percentage [0f, 1f] of animation progress for this confetto.
+     * @param canvas             the canvas to draw on.
```
Comment (jinatonic) [inline confetti/src/main/java/com/github/jinatonic/confetti/confetto/Confetto.java]: revert these changes

File: confetti/src/main/java/com/github/jinatonic/confetti/confetto/Confetto.java [hunk 3]
```diff
@@ -70,6 +65,77 @@
     private float overrideX, overrideY;
     private float overrideDeltaX, overrideDeltaY;
 
+    // Visible for testing
+    protected static Long computeMillisToReachTarget(Float targetVelocity, float initialVelocity,
```
Comment (jinatonic) [inline confetti/src/main/java/com/github/jinatonic/confetti/confetto/Confetto.java]: this is inconsistent with the style of rest of the repo, use double tabs for overflow lines.

Comment (jinatonic) [reply confetti/src/main/java/com/github/jinatonic/confetti/confetto/Confetto.java]: this method is also not changed, please revert the change and move it back to its original location. don't want to unnecessarily modify git blame.

File: confetti/src/main/java/com/github/jinatonic/confetti/confetto/Confetto.java [hunk 4]
```diff
@@ -70,6 +65,77 @@
...
+                    return null;
+                }
+            }
+        } else {
+            return null;
+        }
+    }
+
+    // Visible for testing
+    protected static long computeBound(float initialPos, float velocity, float acceleration,
```
Comment (jinatonic) [inline confetti/src/main/java/com/github/jinatonic/confetti/confetto/Confetto.java]: same thing for this, please do not move methods around unnecessarily, also makes it harder to review.

File: confetti/src/main/java/com/github/jinatonic/confetti/confetto/Confetto.java [hunk 5]
```diff
@@ -281,13 +278,23 @@ public boolean applyUpdate(long passedTime) {
         startedAnimation = animatedTime >= 0;
 
         if (startedAnimation && !terminated) {
-            currentX = computeDistance(animatedTime, initialX, initialVelocityX, accelerationX,
+            float[] pairs = new float[2];
```
Comment (jinatonic) [inline confetti/src/main/java/com/github/jinatonic/confetti/confetto/Confetto.java]: move this into instance variable to avoid allocating this array on every draw, call it `workPairs` or something.

File: confetti/src/main/java/com/github/jinatonic/confetti/confetto/Confetto.java [hunk 6]
```diff
@@ -304,18 +311,24 @@ public boolean applyUpdate(long passedTime) {
         return !terminated;
     }
 
-    private float computeDistance(long t, float xi, float vi, float ai, Long targetTime,
-            Float vTarget) {
+    private void computeDistance(float[] pair, long t, float xi, float vi, float ai, Long targetTime, Float vTarget) {
         if (targetTime == null || t < targetTime) {
-            // distance covered with linear acceleration
+            // velocity with constant acceleration
+            float vX = ai * t + vi;
```
Comment (jinatonic) [inline confetti/src/main/java/com/github/jinatonic/confetti/confetto/Confetto.java]: vX is the same for both if and else conditions, move out of the if-blocks

File: confetti/src/main/java/com/github/jinatonic/confetti/confetto/Confetto.java [hunk 7]
```diff
@@ -304,18 +311,24 @@ public boolean applyUpdate(long passedTime) {
         return !terminated;
     }
 
-    private float computeDistance(long t, float xi, float vi, float ai, Long targetTime,
-            Float vTarget) {
+    private void computeDistance(float[] pair, long t, float xi, float vi, float ai, Long targetTime, Float vTarget) {
```
Comment (jinatonic) [inline confetti/src/main/java/com/github/jinatonic/confetti/confetto/Confetto.java]: we follow a line limit of 100 throughout this repo

File: confetti/src/main/java/com/github/jinatonic/confetti/confetto/Confetto.java [hunk 8]
```diff
@@ -326,10 +339,12 @@ private float computeDistance(long t, float xi, float vi, float ai, Long targetT
      */
     public void draw(Canvas canvas) {
         if (touchOverride) {
-            draw(canvas, overrideX + overrideDeltaX, overrideY + overrideDeltaY, currentRotation,
-                    percentageAnimated);
+            velocityTracker.computeCurrentVelocity(1);
```
Comment (jinatonic) [inline confetti/src/main/java/com/github/jinatonic/confetti/confetto/Confetto.java]: not a fan of using velocityTracker to recompute velocity on every draw after the user touches the confetto, it can be a potentially expensive process. the point of velocity tracker is to track the user's touch and override the initial drawing states to take the user's motion into account. the code inside `computeDistance` SHOULD just work since it uses the overridden state variables, but please double check to make sure.

**Review by jinatonic — CHANGES_REQUESTED** [review]

**Comment by magneticflux-** [comment]
I applied your suggestions and will investigate the usage of `velocityTracker` to see if I can remove the velocity computation.

**Comment by magneticflux-** [comment]
After some profiling of the `draw` method, I have these results:

/ | `computeCurrentVelocity` | `getXVelocity` | `getYVelocity`
--- | --- | --- | ---
Wall clock time: | 73µs (`getPrivateField`) + 47µs (`computeCurrentVelocity`) = 120µs | 90µs (`getPrivateField`) + 26µs (`getXVelocity`) + 93µs (`setPrivateField`) = 209µs | 90µs (`getPrivateField`) + 26µs (`getXVelocity`) + 93µs (`setPrivateField`) = 209µs

As it turns out, the time spent computing the velocity in a native method is dwarfed by the time spent merely accessing a private field. Perhaps making all fields public would net a greater performance gain?

Without using private fields:

/ | `computeCurrentVelocity` | `getXVelocity` | `getYVelocity`
--- | --- | --- | ---
Wall clock time: | 47µs (`computeCurrentVelocity`) = 47µs | 26µs (`getXVelocity`) + 93µs (`setPrivateField`) = 119µs | 26µs (`getXVelocity`) + 93µs (`setPrivateField`) = 119µs

**Comment by jinatonic** [comment]
thats interesting, and it looks like both are insignificant overall so no point in prematurely optimizing (definitely don't want to make the fields public just to gain a few microseconds. but i still think setting velocity two different ways (one from velocitytracker and another from raw computation) is confusing and unnecessary. my original point with regarding to `computeDistance` should properly compute the velocity after the user drags still stands. can you verify if the code is correct without going through velocitytracker?

**Comment by magneticflux-** [comment]
While the velocity is correct without using the `VelocityTracker` both before and after the user touches a confetto, *during* the touch, the velocity is not updated. If a line is drawn in the confetto based on the expected displacement, this results in the line not changing while a user is touching it.

However, if the `VelocityTracker` *is* used, the velocity is updated while the user is touching it. I don't need touch functionality for my use case, but someone in the future might use these two features together and run into unexpected results.

**Comment by magneticflux-** [comment]
In 3b52a1d I tried to restructure the usage of VelocityTracker to make it easier to understand. Now, all usages of the `velocityTracker` field occur in the `onTouch.*` methods and modify `override.*X` and `override.*Y` fields, like how it was originally with `override(X|Y)`.

**Comment by magneticflux-** [comment]
@jinatonic Do you have any other requests for changes or discussion before this is ready to merge?

**Review by jinatonic — APPROVED** [review]
sorry for the super late approve, this totally fell off of my radar.
